src/wasm2.rs

```diff
@@ -0,0 +1,222 @@
+        messages.retain(|_x| false);
+        assert!(messages.len() == 0);
+        async_std::task::Poll::Ready(Ok(byte_count))
+    }
+}
+
+struct Outgoing {
+    server: Arc<Lsp>,
+    messages: Arc<Mutex<Vec<String>>>,
+}
+
+impl Outgoing {
+    pub fn new(server: Arc<Lsp>) -> Self {
+        Outgoing {
+            server,
+            messages: Arc::new(Mutex::new(vec![])),
+        }
+    }
+}
+
+impl async_std::io::Write for Outgoing {
+    #[inline]
+    fn poll_write(
+        self: std::pin::Pin<&mut Self>,
+        _: &mut async_std::task::Context<'_>,
+        buf: &[u8],
+    ) -> async_std::task::Poll<async_std::io::Result<usize>> {
+        let string = std::str::from_utf8(buf).unwrap();
+        log::debug!("string: \"{}\"", string);
+        let parts = string.split("Content-Length:");
+        let mut byte_count = 0;
+        for part in parts.filter(|msg| !msg.is_empty()) {
+            let message = format!("Content-Length:{}", part);
+            byte_count += message.len();
+            self.server.fire(&message);
+        }
+
+        async_std::task::Poll::Ready(Ok(byte_count))
+    }
+
+    #[inline]
+    fn poll_flush(
+        self: std::pin::Pin<&mut Self>,
+        _: &mut async_std::task::Context<'_>,
+    ) -> async_std::task::Poll<async_std::io::Result<()>> {
+        async_std::task::Poll::Ready(Ok(()))
+    }
+
+    #[inline]
+    fn poll_close(
+        self: std::pin::Pin<&mut Self>,
+        _: &mut async_std::task::Context<'_>,
+    ) -> async_std::task::Poll<async_std::io::Result<()>> {
+        async_std::task::Poll::Ready(Ok(()))
+    }
+}
+
+/// Lsp is the core lsp server interface.
+#[wasm_bindgen]
+pub struct Lsp {
+    message_handlers: Vec<js_sys::Function>,
+    incoming: Arc<Mutex<Vec<String>>>,
+}
+
+impl Default for Lsp {
+    fn default() -> Self {
+        console_error_panic_hook::set_once();
+
+        let incoming = Arc::new(Mutex::new(vec![]));
+
+        Lsp {
+            message_handlers: vec![],
+            incoming,
+        }
+    }
+}
+
+#[wasm_bindgen]
+impl Lsp {
+    #[wasm_bindgen(constructor)]
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Fire the message handlers with the server message.
+    pub(crate) fn fire(&self, msg: &str) {
+        for handler in self.message_handlers.iter() {
+            // Set the context to `undefined` explicitly, so the error
+            // message on `this` usage is clear.
+            log::debug!("firing");
+            if let Err(err) = handler.call1(
+                &JsValue::UNDEFINED,
+                &msg.into()
+            ) {
+                log::error!("{:?}", err);
+            }
+        }
+    }
+
+    /// Attach a message handler to server messages.
+    #[allow(non_snake_case)]
+    pub fn onMessage(&mut self, func: js_sys::Function) {
+        self.message_handlers.push(func)
+    }
+
+    /// Send a message to the server.
+    pub fn send(&mut self, msg: String) -> js_sys::Promise {
```

Comment:
This could probably just return `()` now, no?

Reply:
As it acquires a mutex lock, it needs to return a promise. I learned this the hard way.

Reply:
That is not something I would have expected, weird.